src/coreComponents/constitutive/fluid/multifluid/compositional/CompositionalMultiphaseFluidUpdates.hpp

```diff
@@ -0,0 +1,431 @@
+                                     MultiFluidBase::PhaseProp::ViewType phaseVisc,
+                                     MultiFluidBase::PhaseProp::ViewType phaseEnthalpy,
+                                     MultiFluidBase::PhaseProp::ViewType phaseInternalEnergy,
+                                     MultiFluidBase::PhaseComp::ViewType phaseCompFrac,
+                                     MultiFluidBase::FluidProp::ViewType totalDensity ):
+  MultiFluidBase::KernelWrapper( componentMolarWeight,
+                                 useMass,
+                                 std::move( phaseFrac ),
+                                 std::move( phaseDens ),
+                                 std::move( phaseMassDensity ),
+                                 std::move( phaseVisc ),
+                                 std::move( phaseEnthalpy ),
+                                 std::move( phaseInternalEnergy ),
+                                 std::move( phaseCompFrac ),
+                                 std::move( totalDensity ) ),
+  m_componentProperties( componentProperties.createKernelWrapper() ),
+  m_flash( flash.createKernelWrapper() ),
+  m_phase1( phase1.createKernelWrapper()),
+  m_phase2( phase2.createKernelWrapper()),
+  m_phase3( phase3.createKernelWrapper())
+{}
+
+template< typename FLASH, typename PHASE1, typename PHASE2, typename PHASE3 >
+GEOS_HOST_DEVICE
+GEOS_FORCE_INLINE
+void
+CompositionalMultiphaseFluidUpdates< FLASH, PHASE1, PHASE2, PHASE3 >::compute(
+  real64 const pressure,
+  real64 const temperature,
+  arraySlice1d< real64 const, compflow::USD_COMP - 1 > const & composition,
+  arraySlice1d< real64, multifluid::USD_PHASE - 2 > const & phaseFrac,
+  arraySlice1d< real64, multifluid::USD_PHASE - 2 > const & phaseDens,
+  arraySlice1d< real64, multifluid::USD_PHASE - 2 > const & phaseMassDens,
+  arraySlice1d< real64, multifluid::USD_PHASE - 2 > const & phaseVisc,
+  arraySlice1d< real64, multifluid::USD_PHASE - 2 > const & phaseEnthalpy,
+  arraySlice1d< real64, multifluid::USD_PHASE - 2 > const & phaseInternalEnergy,
+  arraySlice2d< real64, multifluid::USD_PHASE_COMP - 2 > const & phaseCompFrac,
+  real64 & totalDens ) const
+{
+  GEOS_UNUSED_VAR( phaseEnthalpy, phaseInternalEnergy );
+  GEOS_UNUSED_VAR( pressure, temperature );
+
+  integer constexpr maxNumComp = MultiFluidBase::MAX_NUM_COMPONENTS;
+  integer constexpr maxNumPhase = MultiFluidBase::MAX_NUM_PHASES;
+  integer const numComp = numComponents();
+  integer const numPhase = numPhases();
+
+  // 1. Convert input mass fractions to mole fractions and keep derivatives
+
+  stackArray1d< real64, maxNumComp > compMoleFrac( numComp );
+  if( m_useMass )
+  {
+    convertToMoleFractions< maxNumComp >( composition, compMoleFrac );
+  }
+  else
+  {
+    for( integer ic = 0; ic < numComp; ++ic )
+    {
+      compMoleFrac[ic] = composition[ic];
+    }
+  }
+
+  // 2. Compute phase fractions and phase component fractions
+  m_flash.compute( pressure,
```

Comment:
it overall looks ok, following current GEOS practices for fluids
but i still do not fully understand why flash compute and phase computes can't become separate kernels?
would simplify the code imho and avoid zillions of template functions
@rrsettgast @klevzoff @francoishamon need you opinion please

Reply:
maybe i could try to implement what i am talking about for a simple fluid to see how it would look...

Reply:
Yes, a dummy design would be helpful.  For the moment I suggest we keep this PR moving forward so we can get baselines in the code, but we can always revisit the design.

Reply:
The original design was influenced primarily by `PVTPackage` which provided a single entry point for all calculations. This does not need to be the case any more.

However a secondary reason concerns performance, especially on GPUs. Split kernel design introduces additional latency and requires input data to be brought from DRAM into SM registers multiple times. Fused kernels are typically much more efficient in that sense and also allow intermediate computation results can be reused in-kernel without storing them into memory.

In addition, our current approach to running these kernels is to have the solver do it rather than the model. This allows fluid compute to be fused into another kernel with a single level of model dispatch, as is the case with Dirichlet boundary conditions in `CompositionalMultiphaseFVM`, which avoids the need to allocate memory to store all fluid properties at those BC locations. I don't see an easy way to do this with split kernel approach, but I'm open to experiments.

Reply:
I will presently carry on as is and welcome ideas on how to rework this.